crates/node/src/roles/base.rs

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2024 -  Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::pin::Pin;
+
+use anyhow::Context;
+use futures::{Stream, StreamExt};
+
+use restate_core::{
+    cancellation_watcher,
+    network::{Incoming, MessageRouterBuilder, NetworkError},
+    task_center,
+    worker_api::ProcessorsManagerHandle,
+    ShutdownError, TaskKind,
+};
+use restate_types::net::node::{GetNodeState, NodeStateResponse};
+
+pub struct BaseRole {
```

Comment:
Question for all: how should we think about putting responsibilities here vs. in the gRPC `NodeSvc` service? When is one a better choice than the other?

Naming: is this really a role, or just a baseline `NodeService`?

Reply:
based on my understanding of the node networking. the BaseRole can serve peer to peer communication across nodes, which is powered by the node core `networking`. In this model messages can be exchanged knowing only the NodeID.

A gRPC service like NodeSvc on the other hand can serve external clients requests (like restatectl) to answer questions like GetIdent, ClusterState, etc... and in this model you need to know which IP and port the service is listening